backend/app/db/dataset/version.py

```diff
@@ -165,3 +173,298 @@ async def _freeze_dataset_visualization(
+                    ThumbnailFreezeDB.origin_id == PydanticObjectId(thumbnail.id)
+                )
+                is None
+            ):
+                fs.remove_object(settings.MINIO_BUCKET_NAME, str(thumbnail.id))
+
+        return thumbnail.dict()
+
+
+async def _delete_file(file, fs):
+    # delete mongo
+    await file.delete()
+
+    if fs is not None:
+        # delete file raw bytes if not used anywhere else
+        if (
+            await FileFreezeDB.find_one(
+                FileFreezeDB.origin_id == PydanticObjectId(file.id)
+            )
+            is None
+        ):
+            fs.remove_object(settings.MINIO_BUCKET_NAME, str(file.id))
+
+    return file.dict()
+
+
+async def _delete_frozen_file(frozen_file, fs):
+    # delete mongo
+    await frozen_file.delete()
+
+    if fs is not None:
+        # delete file raw bytes if not used anywhere else, check both frozen and well as current
+        if (
+            await FileFreezeDB.find_one(
+                FileFreezeDB.origin_id == PydanticObjectId(frozen_file.origin_id)
+            )
+            is None
+        ) and (
+            await FileDB.find_one(FileDB.id == PydanticObjectId(frozen_file.origin_id))
+        ) is None:
+            fs.remove_object(settings.MINIO_BUCKET_NAME, str(frozen_file.origin_id))
+
+    return frozen_file.dict()
+
+
+async def _delete_frozen_dataset(frozen_dataset, fs, hard_delete=False):
+    # delete metadata
+    await MetadataFreezeDB.find(
+        MetadataFreezeDB.resource.resource_id == PydanticObjectId(frozen_dataset.id),
+        MetadataFreezeDB.resource.collection == "datasets",
+    ).delete()
+
+    # delete folders
+    await FolderFreezeDB.find(
+        FolderFreezeDB.dataset_id == PydanticObjectId(frozen_dataset.id)
+    ).delete()
+
+    # delete dataset thumbnails
+    await _delete_frozen_thumbnail(frozen_dataset, fs)
+
+    # delete dataset visualization
+    await _delete_frozen_visualizations(frozen_dataset, fs)
+
+    # delete files and file associate resources
+    async for frozen_file in FileFreezeDB.find(
+        FileFreezeDB.dataset_id == PydanticObjectId(frozen_dataset.id)
+    ):
+        if frozen_file.storage_type == StorageType.LOCAL:
+            await remove_frozen_local_file_entry(frozen_file.id)
+        else:
+            await remove_frozen_file_entry(frozen_file.id, fs)
+
+    # delete authorizations
+    await AuthorizationDB.find(
+        AuthorizationDB.dataset_id == PydanticObjectId(frozen_dataset.id)
+    ).delete()
+
+    # If all above succeeded
+    if hard_delete:
+        await frozen_dataset.delete()
+    else:
+        frozen_dataset.deleted = True
+        await frozen_dataset.save()
+
+    return frozen_dataset.dict()
+
+
+# TODO: Move this to MongoDB middle layer
+async def remove_file_entry(
+    file_id: Union[str, ObjectId],
+    fs: Minio,
+    es: Elasticsearch,
+):
+    """Remove FileDB object into MongoDB, Minio, and associated metadata and version information."""
```

Comment:
Can you add the parameters and return type to the doc string? May be add them to all the public functions?

Reply:
Yes, I'm deleting the raw bytes in this private function 
```  
# delete file raw bytes if not used anywhere else
await _delete_file(file, fs)
```

Reply:
I added doc strings to all the functions here.